src/snowflake/snowpark/_internal/ast.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from dataclasses import dataclass
+from typing import Callable, Tuple
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+from snowflake.snowpark.version import VERSION
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+@dataclass
+class TrackedCallable:
+    var_id: int
+    func: Callable
+
+
+class AstBatch:
+    """
+    A batch of AST statements. This class is used to generate AST requests.
+
+    The core statement types are:
+    - Assign: Creates a new variable and assigns a value to it.
+    - Eval: Evaluates a variable.
+    """
+
+    # Function used to generate request IDs. This is overridden in some tests.
+    generate_request_id = uuid.uuid4
+
+    def __init__(self, session) -> None:
+        """
+        Initializes a new AST batch.
+
+        Args:
+            session: The Snowpark session.
+        """
+        self._session = session
+        self.reset_id_gen()
+        self._init_batch()
+
+        # Track callables in this dict (memory id -> TrackedCallable).
+        self._callables = {}
+
+    def reset_id_gen(self):
+        """Resets the ID generator."""
+        self._id_gen = itertools.count(start=1)
+
+    def assign(self, symbol=None):
+        """
+        Creates a new assignment statement.
+
+        Args:
+            symbol: An optional symbol to name the new variable.
+        """
+        stmt = self._request.body.add()
+        # TODO: extended BindingId spec from the branch snowpark-ir.
+        stmt.assign.uid = next(self._id_gen)
```

Comment:
Do we need to make sure AST is thread safe? cc @sfc-gh-aalam @sfc-gh-aling 

If current analyzer is thread safe, then AST code should be thread safe too right?

Reply:
When the thread safety work came up, the discussion I remember was that `_ast_batch` is a field of `Session`, and thus needs to be thread safe similarly to `Session`. If this happens by virtue of methods getting called through APIs already synchronized by the `Session` mutex, nothing to do for us. Otherwise, we can add internal synchronization to `AstBatch`. I just need to know what the contract needs to be.

Reply:
make sense. Let's move forward since no concern here.
Also, let's add one line comment to describe `AstBatch` is not thread safe for now.

Reply:
Done.